File: python/bc-components/src/bc_components/encrypted_key/_key_derivation_params.py
"""Union type for key derivation parameters."""

from __future__ import annotations

from bc_tags import CBOR

from ..symmetric._encrypted_message import EncryptedMessage
from ..symmetric._symmetric_key import SymmetricKey
from ._argon2id_params import Argon2idParams
from ._hkdf_params import HKDFParams
from ._key_derivation_method import KeyDerivationMethod
from ._pbkdf2_params import PBKDF2Params
from ._scrypt_params import ScryptParams


class KeyDerivationParams:
    """Union of all key derivation parameter types.

    Wraps one of HKDFParams, PBKDF2Params, ScryptParams, or Argon2idParams,
    providing a unified interface for lock/unlock operations.
    """

    __slots__ = ("_inner",)

    def __init__(
        self,
        inner: HKDFParams | PBKDF2Params | ScryptParams | Argon2idParams,
    ) -> None:
        self._inner = inner

    # --- Construction ---

    @staticmethod
    def hkdf(params: HKDFParams) -> KeyDerivationParams:
        return KeyDerivationParams(params)

    @staticmethod
    def pbkdf2(params: PBKDF2Params) -> KeyDerivationParams:
        return KeyDerivationParams(params)

    @staticmethod
    def scrypt(params: ScryptParams) -> KeyDerivationParams:
        return KeyDerivationParams(params)

    @staticmethod
    def argon2id(params: Argon2idParams) -> KeyDerivationParams:
        return KeyDerivationParams(params)

    # --- Properties ---

    @property
    def inner(self) -> HKDFParams | PBKDF2Params | ScryptParams | Argon2idParams:
        return self._inner

    def method(self) -> KeyDerivationMethod:
        """Return the key derivation method associated with these parameters."""
        if isinstance(self._inner, HKDFParams):
            return KeyDerivationMethod.HKDF
        elif isinstance(self._inner, PBKDF2Params):
            return KeyDerivationMethod.PBKDF2
        elif isinstance(self._inner, ScryptParams):
            return KeyDerivationMethod.SCRYPT
        elif isinstance(self._inner, Argon2idParams):
            return KeyDerivationMethod.ARGON2ID
        raise ValueError("Unknown inner params type")  # pragma: no cover

    def is_password_based(self) -> bool:
        """Return True if the derivation method is password-based."""
        return isinstance(
            self._inner, (PBKDF2Params, ScryptParams, Argon2idParams)
        )

    # --- Key derivation ---

    def lock(
        self,
        content_key: SymmetricKey,
        secret: bytes | bytearray,
    ) -> EncryptedMessage:
        """Derive a key from *secret* and encrypt *content_key*."""
        return self._inner.lock(content_key, secret)

    # --- Display ---

    def __str__(self) -> str:
        return str(self._inner)

    def __repr__(self) -> str:
        return f"KeyDerivationParams({self._inner!r})"

    def __eq__(self, other: object) -> bool:
        if isinstance(other, KeyDerivationParams):
            return self._inner == other._inner
        return NotImplemented

    def __hash__(self) -> int:
        return hash(self._inner)

    # --- CBOR ---

    def to_cbor(self) -> CBOR:
        """Encode the inner params as CBOR."""
        return self._inner.to_cbor()

    @staticmethod
    def from_cbor(cbor: CBOR) -> KeyDerivationParams:
        """Decode from CBOR by inspecting the first array element (method index)."""
        a = cbor.try_array()
        if not a:
            raise ValueError("KeyDerivationParams: empty array")
        index = a[0].try_int()
        method = KeyDerivationMethod.from_index(index)
        if method is None:
            raise ValueError(f"Invalid KeyDerivationMethod index: {index}")
        if method == KeyDerivationMethod.HKDF:
            return KeyDerivationParams(HKDFParams.from_cbor(cbor))
        elif method == KeyDerivationMethod.PBKDF2:
            return KeyDerivationParams(PBKDF2Params.from_cbor(cbor))
        elif method == KeyDerivationMethod.SCRYPT:
            return KeyDerivationParams(ScryptParams.from_cbor(cbor))
        elif method == KeyDerivationMethod.ARGON2ID:
            return KeyDerivationParams(Argon2idParams.from_cbor(cbor))
        raise ValueError(f"Unsupported KeyDerivationMethod: {method}")  # pragma: no cover

File: typescript/bc-components/src/encrypted-key/key-derivation-params.ts
import { type Cbor } from '@bc/dcbor';

import type { EncryptedMessage } from '../symmetric/encrypted-message.js';
import type { SymmetricKey } from '../symmetric/symmetric-key.js';
import { Argon2idParams } from './argon2id-params.js';
import { HKDFParams } from './hkdf-params.js';
import { KeyDerivationMethod, keyDerivationMethodFromIndex } from './key-derivation-method.js';
import { PBKDF2Params } from './pbkdf2-params.js';
import { ScryptParams } from './scrypt-params.js';
import { BCComponentsError } from '../error.js';

type KeyDerivationParamsVariant =
    | { type: 'hkdf'; params: HKDFParams }
    | { type: 'pbkdf2'; params: PBKDF2Params }
    | { type: 'scrypt'; params: ScryptParams }
    | { type: 'argon2id'; params: Argon2idParams };

export class KeyDerivationParams {
    readonly #variant: KeyDerivationParamsVariant;

    private constructor(variant: KeyDerivationParamsVariant) {
        this.#variant = variant;
    }

    static hkdf(params: HKDFParams): KeyDerivationParams {
        return new KeyDerivationParams({ type: 'hkdf', params });
    }

    static pbkdf2(params: PBKDF2Params): KeyDerivationParams {
        return new KeyDerivationParams({ type: 'pbkdf2', params });
    }

    static scrypt(params: ScryptParams): KeyDerivationParams {
        return new KeyDerivationParams({ type: 'scrypt', params });
    }

    static argon2id(params: Argon2idParams): KeyDerivationParams {
        return new KeyDerivationParams({ type: 'argon2id', params });
    }

    method(): KeyDerivationMethod {
        switch (this.#variant.type) {
            case 'hkdf':
                return KeyDerivationMethod.HKDF;
            case 'pbkdf2':
                return KeyDerivationMethod.PBKDF2;
            case 'scrypt':
                return KeyDerivationMethod.Scrypt;
            case 'argon2id':
                return KeyDerivationMethod.Argon2id;
        }
    }

    isPasswordBased(): boolean {
        return this.#variant.type !== 'hkdf';
    }

    isSshAgent(): boolean {
        return false;
    }

    lock(contentKey: SymmetricKey, secret: Uint8Array): EncryptedMessage {
        switch (this.#variant.type) {
            case 'hkdf':
                return this.#variant.params.lock(contentKey, secret);
            case 'pbkdf2':
                return this.#variant.params.lock(contentKey, secret);
            case 'scrypt':
                return this.#variant.params.lock(contentKey, secret);
            case 'argon2id':
                return this.#variant.params.lock(contentKey, secret);
        }
    }

    unlock(encryptedMessage: EncryptedMessage, secret: Uint8Array): SymmetricKey {
        switch (this.#variant.type) {
            case 'hkdf':
                return this.#variant.params.unlock(encryptedMessage, secret);
            case 'pbkdf2':
                return this.#variant.params.unlock(encryptedMessage, secret);
            case 'scrypt':
                return this.#variant.params.unlock(encryptedMessage, secret);
            case 'argon2id':
                return this.#variant.params.unlock(encryptedMessage, secret);
        }
    }

    toCbor(): Cbor {
        switch (this.#variant.type) {
            case 'hkdf':
                return this.#variant.params.toCbor();
            case 'pbkdf2':
                return this.#variant.params.toCbor();
            case 'scrypt':
                return this.#variant.params.toCbor();
            case 'argon2id':
                return this.#variant.params.toCbor();
        }
    }

    static fromCbor(cbor: Cbor): KeyDerivationParams {
        const a = cbor.toArray();
        const index = Number(a[0]!.toInteger());
        const method = keyDerivationMethodFromIndex(index);
        switch (method) {
            case KeyDerivationMethod.HKDF:
                return KeyDerivationParams.hkdf(HKDFParams.fromCbor(cbor));
            case KeyDerivationMethod.PBKDF2:
                return KeyDerivationParams.pbkdf2(PBKDF2Params.fromCbor(cbor));
            case KeyDerivationMethod.Scrypt:
                return KeyDerivationParams.scrypt(ScryptParams.fromCbor(cbor));
            case KeyDerivationMethod.Argon2id:
                return KeyDerivationParams.argon2id(Argon2idParams.fromCbor(cbor));
            default:
                throw BCComponentsError.general('Invalid KeyDerivationMethod');
        }
    }

    toString(): string {
        switch (this.#variant.type) {
            case 'hkdf':
                return this.#variant.params.toString();
            case 'pbkdf2':
                return this.#variant.params.toString();
            case 'scrypt':
                return this.#variant.params.toString();
            case 'argon2id':
                return this.#variant.params.toString();
        }
    }
}
